AX: Support multiple replacement strings in AXTextOperation

Sameesunkaria opened the pull request:
#### f20ec536521f2da91bd2993b1a9034e456d1f907
<pre>
AX: Support multiple replacement strings in AXTextOperation
<a href="https://bugs.webkit.org/show_bug.cgi?id=279888">https://bugs.webkit.org/show_bug.cgi?id=279888</a>
<a href="https://rdar.apple.com/problem/136220931">rdar://problem/136220931</a>

Reviewed by Tyler Wilcock.

WebKit’s `AXTextOperation` parameterized accessibility attribute, among its other capabilities,
lets you replace text across multiple text marker ranges in an editable text field. It takes a
dictionary as its parameter, with an array of text marker ranges to be replaced, and one
replacement string. The existing API mirrors the contract for a simple find-and-replace
operation. Utilizing `AXSearchTextWithCriteria`, you can locate text marker ranges for a search
string and, substitute them with a replacement string using `AXTextOperation`. However,
AXTextOperation’s usability is severely limited beyond this specific use case.

This change adds the ability to replace _individual_ ranges with different replacement strings,
allowing AXTextOperations to be used outside the narrow scope of a simple find-and-replace
operation, enabling richer replacements from an accessibility client, like Grammarly (which
intends to use this attribute).

A new `AXTextOperationIndividualReplacementStrings` key has been added to the `AXTextOperation`
parameter dictionary, which expects an array of strings as its value. This array is expected to
have the same length as the `AXTextOperationMarkerRanges` array, where each replacement string
corresponds to one replacement range.

Client pseudo-code:

```
// Editable text: Lorem ipsum dolor sit amet, consectetur adipiscing elit.
// To replace &quot;dolor&quot; -&gt; &quot;foo&quot; and &quot;elit&quot; -&gt; &quot;bar&quot;:
CFTypeRef result;
AXUIElementCopyParameterizedAttributeValue(element, &quot;AXTextOperation&quot;, @[
  NSAccessibilityTextOperationType: NSAccessibilityTextOperationReplace,
  NSAccessibilityTextOperationIndividualReplacementStrings: @[ @&quot;bar&quot;, @&quot;foo&quot; ],
  NSAccessibilityTextOperationMarkerRanges: @[ rangeOfElit, rangeOfDolor ]
], &amp;result)
```

This is an additive change to the AXTextOperation API. The existing
`AXTextOperationReplacementString` key for providing a single replacement string will continue
to work as expected.

* LayoutTests/accessibility/mac/text-operation/text-operation-replace-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case.html:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace.html:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::performTextOperation):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(accessibilityTextCriteriaForParameterizedAttribute):
(accessibilityTextOperationForParameterizedAttribute):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::performTextOperation):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::textOperationParameterizedAttribute):
(WTR::AccessibilityUIElement::performTextOperation):

Canonical link: <a href="https://commits.webkit.org/284044@main">https://commits.webkit.org/284044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d5a24ba4b68fda677583b9cabdfd8bee8fa4ede9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68025 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47417 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20684 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72084 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19170 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70142 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18986 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/54379 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/12783 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71092 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43424 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58787 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34840 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40093 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16203 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/17527 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62068 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16544 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73783 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11996 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15823 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61829 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12034 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58862 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61844 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15148 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9743 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3357 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43217 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44291 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45486 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44033 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case.html:
```diff
@@ -52,6 +53,18 @@
+        output += expect("operationResult[1]", "'Two'");
+        output += expect("operationResult[2]", "'One'");
+        output += expect("text.stringValue.replace(/\\s/g, ' ')", "'AXValue: TEXT4: One quick brown Two jumps over the Three dog.'");
+
```
These new test cases are great! Another nice one to have would be a case where there are more markers than replacement strings.

Sameesunkaria replied:
That's a good idea. I've added the test case :)

twilco reviewed (commented):
Looks great! Only two minor comments.

Could you please update your commit message and PR description to mention that Grammarly intends to use this new attribute? Sometimes we audit our public API, and if they aren't used by Apple-internal clients, and not used by open source third-party clients, we consider removing them. So having this intent in the git history will help us know to keep this API around.

Sameesunkaria committed:
AX: Support multiple replacement strings in AXTextOperation
https://bugs.webkit.org/show_bug.cgi?id=279888
rdar://problem/136220931

Reviewed by Tyler Wilcock.

WebKit’s `AXTextOperation` parameterized accessibility attribute, among its other capabilities,
lets you replace text across multiple text marker ranges in an editable text field. It takes a
dictionary as its parameter, with an array of text marker ranges to be replaced, and one
replacement string. The existing API mirrors the contract for a simple find-and-replace
operation. Utilizing `AXSearchTextWithCriteria`, you can locate text marker ranges for a search
string and, substitute them with a replacement string using `AXTextOperation`. However,
AXTextOperation’s usability is severely limited beyond this specific use case.

This change adds the ability to replace _individual_ ranges with different replacement strings,
allowing AXTextOperations to be used outside the narrow scope of a simple find-and-replace
operation, enabling richer replacements from an accessibility client, like Grammarly (which
intends to use this attribute).

A new `AXTextOperationIndividualReplacementStrings` key has been added to the `AXTextOperation`
parameter dictionary, which expects an array of strings as its value. This array is expected to
have the same length as the `AXTextOperationMarkerRanges` array, where each replacement string
corresponds to one replacement range.

Client pseudo-code:

```
// Editable text: Lorem ipsum dolor sit amet, consectetur adipiscing elit.
// To replace "dolor" -> "foo" and "elit" -> "bar":
CFTypeRef result;
AXUIElementCopyParameterizedAttributeValue(element, "AXTextOperation", @[
  NSAccessibilityTextOperationType: NSAccessibilityTextOperationReplace,
  NSAccessibilityTextOperationIndividualReplacementStrings: @[ @"bar", @"foo" ],
  NSAccessibilityTextOperationMarkerRanges: @[ rangeOfElit, rangeOfDolor ]
], &result)
```

This is an additive change to the AXTextOperation API. The existing
`AXTextOperationReplacementString` key for providing a single replacement string will continue
to work as expected.

* LayoutTests/accessibility/mac/text-operation/text-operation-replace-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case-expected.txt:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace-preserve-case.html:
* LayoutTests/accessibility/mac/text-operation/text-operation-replace.html:
* Source/WebCore/accessibility/AXCoreObject.h:
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::performTextOperation):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:
(accessibilityTextCriteriaForParameterizedAttribute):
(accessibilityTextOperationForParameterizedAttribute):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.cpp:
(WTR::AccessibilityUIElement::performTextOperation):
* Tools/WebKitTestRunner/InjectedBundle/AccessibilityUIElement.h:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/AccessibilityUIElement.idl:
* Tools/WebKitTestRunner/InjectedBundle/mac/AccessibilityUIElementMac.mm:
(WTR::textOperationParameterizedAttribute):
(WTR::AccessibilityUIElement::performTextOperation):

Canonical link: https://commits.webkit.org/284044@main
